// @urql/exchange-retry/dist/urql-exchange-retry.mjs
import { share as r, makeSubject as e, mergeMap as t, filter as n, takeUntil as a, delay as o, fromValue as i, merge as u } from "wonka";

import { makeOperation as c } from "@urql/core";

function _extends() {
  return (_extends = Object.assign || function(r) {
    for (var e = 1; e < arguments.length; e++) {
      var t = arguments[e];
      for (var n in t) {
        if (Object.prototype.hasOwnProperty.call(t, n)) {
          r[n] = t[n];
        }
      }
    }
    return r;
  }).apply(this, arguments);
}

function retryExchange(s) {
  var y = s.initialDelayMs || 1e3;
  var p = s.maxDelayMs || 15e3;
  var f = s.maxNumberAttempts || 2;
  var v = s.randomDelay || !0;
  var d = s.retryIf || function(r) {
    return r && r.networkError;
  };
  return function(s) {
    var m = s.forward;
    var h = s.dispatchDebug;
    return function(s) {
      var x = r(s);
      var l = e();
      var g = l.source;
      var E = l.next;
      var b = t((function(r) {
        var e = r.key;
        var t = r.context;
        var u = (t.retryCount || 0) + 1;
        var s = t.retryDelay || y;
        var d = Math.random() + 1.5;
        if (v && s * d < p) {
          s *= d;
        }
        var m = n((function(r) {
          return ("query" === r.kind || "teardown" === r.kind) && r.key === e;
        }))(x);
        "production" !== process.env.NODE_ENV && h({
          type: "retryAttempt",
          message: "The operation has failed and a retry has been triggered (" + u + " / " + f + ")",
          operation: r,
          data: {
            retryCount: u
          },
          source: "retryExchange"
        });
        return a(m)(o(s)(i(c(r.kind, r, _extends({}, r.context, {
          retryDelay: s,
          retryCount: u
        })))));
      }))(g);
      return n((function(r) {
        if (!r.error || !d(r.error, r.operation)) {
          return !0;
        }
        if (!((r.operation.context.retryCount || 0) >= f - 1)) {
          E(r.operation);
          return !1;
        }
        "production" !== process.env.NODE_ENV && h({
          type: "retryExhausted",
          message: "Maximum number of retries has been reached. No further retries will be performed.",
          operation: r.operation,
          source: "retryExchange"
        });
        return !0;
      }))(r(m(u([ x, b ]))));
    };
  };
}

export { retryExchange };
