// @ceramicnetwork/streamid/lib/stream-id.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, privateMap, value) {
    if (!privateMap.has(receiver)) {
        throw new TypeError("attempted to set private field on non-instance");
    }
    privateMap.set(receiver, value);
    return value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, privateMap) {
    if (!privateMap.has(receiver)) {
        throw new TypeError("attempted to get private field on non-instance");
    }
    return privateMap.get(receiver);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var _type, _cid;
Object.defineProperty(exports, "__esModule", { value: true });
exports.StreamID = void 0;
const cids_1 = __importDefault(require("cids"));
const multibase_1 = __importDefault(require("multibase"));
const varint_1 = __importDefault(require("varint"));
const concat_1 = __importDefault(require("uint8arrays/concat"));
const to_string_1 = __importDefault(require("uint8arrays/to-string"));
const constants_1 = require("./constants");
const reading_bytes_1 = require("./reading-bytes");
const typescript_memoize_1 = require("typescript-memoize");
const commit_id_1 = require("./commit-id");
const stream_type_1 = require("./stream-type");
function fromBytes(bytes) {
    const [streamCodec, streamCodecRemainder] = reading_bytes_1.readVarint(bytes);
    if (streamCodec !== constants_1.STREAMID_CODEC)
        throw new Error('fromBytes: invalid streamid, does not include streamid codec');
    const [type, streamTypeRemainder] = reading_bytes_1.readVarint(streamCodecRemainder);
    const [cid, cidRemainder] = reading_bytes_1.readCid(streamTypeRemainder);
    if (cidRemainder.length > 0) {
        throw new Error(`Invalid StreamID: contains commit`);
    }
    return new StreamID(type, cid);
}
function fromString(input) {
    const protocolFree = input.replace('ceramic://', '').replace('/ceramic/', '');
    const commitFree = protocolFree.includes('commit') ? protocolFree.split('?')[0] : protocolFree;
    const bytes = multibase_1.default.decode(commitFree);
    return fromBytes(bytes);
}
const TAG = Symbol.for('@ceramicnetwork/streamid/StreamID');
class StreamID {
    constructor(type, cid) {
        this._tag = TAG;
        _type.set(this, void 0);
        _cid.set(this, void 0);
        if (!(type || type === 0))
            throw new Error('constructor: type required');
        if (!cid)
            throw new Error('constructor: cid required');
        __classPrivateFieldSet(this, _type, typeof type === 'string' ? stream_type_1.StreamType.codeByName(type) : type);
        __classPrivateFieldSet(this, _cid, typeof cid === 'string' ? new cids_1.default(cid) : cid);
    }
    static isInstance(instance) {
        return typeof instance === 'object' && '_tag' in instance && instance._tag === TAG;
    }
    get type() {
        return __classPrivateFieldGet(this, _type);
    }
    get typeName() {
        return stream_type_1.StreamType.nameByCode(__classPrivateFieldGet(this, _type));
    }
    get cid() {
        return __classPrivateFieldGet(this, _cid);
    }
    get bytes() {
        const codec = varint_1.default.encode(constants_1.STREAMID_CODEC);
        const type = varint_1.default.encode(this.type);
        return concat_1.default([codec, type, this.cid.bytes]);
    }
    get baseID() {
        return new StreamID(__classPrivateFieldGet(this, _type), __classPrivateFieldGet(this, _cid));
    }
    atCommit(commit) {
        return new commit_id_1.CommitID(__classPrivateFieldGet(this, _type), __classPrivateFieldGet(this, _cid), commit);
    }
    equals(other) {
        if (StreamID.isInstance(other)) {
            return this.type === other.type && this.cid.equals(other.cid);
        }
        else {
            return false;
        }
    }
    toString() {
        return to_string_1.default(multibase_1.default.encode(constants_1.DEFAULT_BASE, this.bytes));
    }
    toUrl() {
        return `ceramic://${this.toString()}`;
    }
    [(_type = new WeakMap(), _cid = new WeakMap(), Symbol.for('nodejs.util.inspect.custom'))]() {
        return `StreamID(${this.toString()})`;
    }
    [Symbol.toPrimitive]() {
        return this.toString();
    }
}
StreamID.fromBytes = fromBytes;
StreamID.fromString = fromString;
__decorate([
    typescript_memoize_1.Memoize(),
    __metadata("design:type", String),
    __metadata("design:paramtypes", [])
], StreamID.prototype, "typeName", null);
__decorate([
    typescript_memoize_1.Memoize(),
    __metadata("design:type", Uint8Array),
    __metadata("design:paramtypes", [])
], StreamID.prototype, "bytes", null);
__decorate([
    typescript_memoize_1.Memoize(),
    __metadata("design:type", StreamID),
    __metadata("design:paramtypes", [])
], StreamID.prototype, "baseID", null);
__decorate([
    typescript_memoize_1.Memoize(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", String)
], StreamID.prototype, "toString", null);
__decorate([
    typescript_memoize_1.Memoize(),
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", String)
], StreamID.prototype, "toUrl", null);
exports.StreamID = StreamID;
