// next-urql/dist/next-urql.js
var e = require("react");

var t = require("react-ssr-prepass");

var n = require("urql");

function _interopDefaultLegacy(e) {
  return e && "object" == typeof e && "default" in e ? e.default : e;
}

var r = _interopDefaultLegacy(e);

var i = _interopDefaultLegacy(t);

function _extends() {
  return (_extends = Object.assign || function(e) {
    for (var t = 1; t < arguments.length; t++) {
      var n = arguments[t];
      for (var r in n) {
        if (Object.prototype.hasOwnProperty.call(n, r)) {
          e[r] = n[r];
        }
      }
    }
    return e;
  }).apply(this, arguments);
}

var a = null;

function resetClient() {
  a = null;
}

function initUrqlClient(e, t) {
  var r = "undefined" == typeof window;
  if (r || !a) {
    (a = n.createClient(_extends({}, e, {
      suspense: t && (r || e.suspense)
    }))).toJSON = function() {
      return null;
    };
  }
  return a;
}

var l;

exports.initUrqlClient = initUrqlClient;

exports.withUrqlClient = function withUrqlClient(t, a) {
  if (!a) {
    a = {};
  }
  return function(o) {
    var u = Boolean((o.getInitialProps || a.ssr) && !a.neverSuspend);
    var WithUrql = function(i) {
      var a = i.pageProps;
      var s = i.urqlClient;
      var c = i.urqlState;
      var f = function objectWithoutProperties(e, t) {
        var n = {};
        for (var r in e) {
          if (Object.prototype.hasOwnProperty.call(e, r) && -1 === t.indexOf(r)) {
            n[r] = e[r];
          }
        }
        return n;
      }(i, [ "pageProps", "urqlClient", "urqlState" ]);
      var p = e.useState(0);
      var v = a && a.urqlState || c;
      var d = r.useMemo((function() {
        if (s) {
          return s;
        }
        if (!l || "undefined" == typeof window) {
          l = n.ssrExchange({
            initialState: v,
            isClient: !0
          });
        } else if (l && "undefined" != typeof window) {
          l.restoreData(v);
        }
        var e = t(l);
        if (!e.exchanges) {
          e.exchanges = [ n.dedupExchange, n.cacheExchange, l, n.fetchExchange ];
        }
        return initUrqlClient(e, u);
      }), [ s, v, p[0] ]);
      return e.createElement(n.Provider, {
        value: d
      }, e.createElement(o, _extends({}, f, {
        pageProps: a,
        urqlClient: d,
        resetUrqlClient: function() {
          resetClient();
          l = n.ssrExchange({
            initialState: void 0
          });
          p[1](p[0] + 1);
        }
      })));
    };
    WithUrql.displayName = "withUrqlClient(" + (o.displayName || o.name || "Component") + ")";
    if (o.getInitialProps || a.ssr) {
      WithUrql.getInitialProps = function(r) {
        try {
          function _temp4() {
            function _temp2() {
              return _extends({}, v, {
                urqlState: c ? c.extractData() : void 0,
                urqlClient: p
              });
            }
            if ("undefined" != typeof window) {
              return _extends({}, v, {
                urqlClient: p
              });
            }
            var t = _extends({}, v, {
              urqlClient: p
            });
            var n = u ? t : {
              pageProps: t
            };
            const r = function() {
              if (!a.neverSuspend) {
                return Promise.resolve(i(e.createElement(l, n))).then((function() {}));
              }
            }();
            return r && r.then ? r.then(_temp2) : _temp2();
          }
          var l = r.AppTree;
          var u = !!r.Component;
          var s = u ? r.ctx : r;
          var c = n.ssrExchange({
            initialState: void 0
          });
          var f = t(c, s);
          if (!f.exchanges) {
            f.exchanges = [ n.dedupExchange, n.cacheExchange, c, n.fetchExchange ];
          }
          var p = initUrqlClient(f, !a.neverSuspend);
          if (p) {
            s.urqlClient = p;
          }
          var v = {};
          const d = function() {
            if (o.getInitialProps) {
              return Promise.resolve(o.getInitialProps(r)).then((function(e) {
                v = e;
              }));
            }
          }();
          return Promise.resolve(d && d.then ? d.then(_temp4) : _temp4());
        } catch (e) {
          return Promise.reject(e);
        }
      };
    }
    return WithUrql;
  };
};
