// storybook-addon-performance/dist/cli/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.printCSVSummary = exports.printCSV = exports.performCalculations = exports.median = exports.combineTaskResultsByGroupId = exports.convertToTaskValueMap = exports.usage = exports.stdout = exports.debug = void 0;
exports.debug = (...args) => console.warn(...args);
exports.stdout = (...args) => console.log(...args);
exports.usage = () => exports.debug(`Usage:
sb-perf <directory> [...<directory>]

Example
sb-perf results-directory > output-file.csv
# OR
sb-perf ABTestDirectory OtherDirectory > output-file.csv
`);
const getTaskValue = (result) => {
    if ('averageMs' in result) {
        return result.averageMs;
    }
    if ('value' in result) {
        return Number(result.value);
    }
    return null;
};
exports.convertToTaskValueMap = (resultMap) => {
    return Object.values(resultMap).reduce((acc, result) => {
        return { ...acc, [result.taskName]: [getTaskValue(result)] };
    }, {});
};
exports.combineTaskResultsByGroupId = (results, [groupId, taskValueMap]) => {
    Object.entries(taskValueMap).forEach(([taskName, value]) => {
        results[groupId] = {
            ...results[groupId],
            [taskName]: results[groupId] && results[groupId][taskName]
                ? results[groupId][taskName].concat(value)
                : value,
        };
    });
    return results;
};
exports.median = (numbers) => {
    const sorted = numbers.slice().sort((a, b) => a - b);
    const middle = Math.floor(sorted.length / 2);
    if (sorted.length % 2 === 0) {
        return (sorted[middle - 1] + sorted[middle]) / 2;
    }
    return sorted[middle];
};
exports.performCalculations = (data) => {
    const numResults = Object.values(data)[0].length;
    return Object.entries(data).map(([key, values]) => ({
        key,
        numResults,
        samples: values.join(','),
        minValue: Math.min(...values),
        maxValue: Math.max(...values),
        meanValue: values.reduce((acc, curr) => acc + curr, 0) / numResults,
        medianValue: exports.median(values),
    }));
};
exports.printCSV = (rows) => {
    const numResults = rows[0].numResults;
    exports.stdout(`type,${Array.from({ length: numResults })
        .map((_, i) => `#${i + 1}`)
        .join(',')},min,max,mean,median`);
    rows.forEach(({ key, samples, minValue, maxValue, meanValue, medianValue }) => {
        exports.stdout(`${key},${samples},${minValue},${maxValue},${meanValue},${medianValue}`);
    });
};
const padded = (padding) => (str) => `${str}`.padEnd(padding);
const paddedKey = padded(50);
const paddedValue = padded(20);
exports.printCSVSummary = (resultNames, results) => {
    exports.debug(`${paddedKey('Type')} | ${resultNames.map(paddedValue).join('| ')}`);
    exports.debug('-'.padEnd(50 + 20 * resultNames.length, '-'));
    exports.stdout(`Type,${resultNames.join(',')}`);
    const toPrint = {};
    results.forEach((rows) => {
        rows.forEach(({ key, medianValue }) => {
            if (toPrint[key]) {
                toPrint[key].push(medianValue);
            }
            else {
                toPrint[key] = [medianValue];
            }
        });
    });
    Object.entries(toPrint).forEach(([key, values]) => {
        exports.debug(`${paddedKey(key)} | ${values.map(paddedValue).join('| ')}`);
        exports.stdout(`${key},${values.join(',')}`);
    });
};
